### PR TITLE
FXIOS-825 ⁃ Add Top Sites & Open Tabs widgets

### PR DESCRIPTION
**Pull request by sblatz**
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Please make sure you've run the test scheme and all the tests pass. If your changes affect existing tests please make sure to update the tests as well.

## Demo

widget work

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXIOS-825)

### CODE REVIEW

**Comment by sblatz** [comment]
Please note I'm currently encountering **one known issue**, that I'd love help on. 

If a user has open tabs, then force quits, the tabs remain in the widget. However, if you click one of the links, it will open it as a new tab, rather than taking you to the one that needs to be restored. I believe this is because I'm using the url within the `sessionData` rather than the one that has the `restore` information (they have like a UUID and a super weird URL). What's odd about this is when I try to open it with the deep link, it fails to restore the tab and just sits there blank. 

Not sure how to best handle that :)

File: Client/Frontend/Browser/TabArchiver.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+              FileManager.default.fileExists(atPath: tabStateArchivePath),
+              let tabData = try? Data(contentsOf: URL(fileURLWithPath: tabStateArchivePath)) else {
+            return [SavedTab]()
+        }
+        
+        let unarchiver = NSKeyedUnarchiver(forReadingWith: tabData)
+        unarchiver.setClass(SavedTab.self, forClassName: "Client.SavedTab")
+        unarchiver.setClass(SessionData.self, forClassName: "Client.SessionData")
+        
+        print("unarchiver decoding...")
```
Comment (sblatz) [inline Client/Frontend/Browser/TabArchiver.swift:20]: Whoops. Totally forgot to clean this file up. I'll do that Monday morning :)

File: package-lock.json [hunk 2]
```diff
@@ -2815,28 +2815,28 @@
       "dependencies": {
         "abbrev": {
           "version": "1.1.1",
-          "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
```
Comment (nbhasin2) [inline package-lock.json:2818]: Not sure why `package-lock.json` needs to change 🤔 
cc @sblatz any thoughts on this?

File: Client/Frontend/Browser/SavedTab+ConfigureExtension.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import Storage
+import Shared
+
+// This cannot be easily imported into extension targets, so we break it out here.
+extension SavedTab {
+    convenience init?(tab: Tab, isSelected: Bool) {
+        assert(Thread.isMainThread)
+        
+        var sessionData = tab.sessionData
+        if sessionData == nil {
```
Comment (nbhasin2) [inline Client/Frontend/Browser/SavedTab+ConfigureExtension.swift:16]: nit: guard statement would be nice

File: Client/Frontend/Browser/SavedTab+ConfigureExtension.swift [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            let currentItem: WKBackForwardListItem! = tab.webView?.backForwardList.currentItem
+
+            // Freshly created web views won't have any history entries at all.
+            // If we have no history, abort.
+            if currentItem != nil {
+                let backList = tab.webView?.backForwardList.backList ?? []
```
Comment (nbhasin2) [inline Client/Frontend/Browser/SavedTab+ConfigureExtension.swift:22]: A single line comment on what back, forward list is trying to do be be nice for any future reader

File: Client/Frontend/Browser/TabManager.swift [hunk 5]
```diff
@@ -196,7 +204,7 @@ class TabManager: NSObject {
         }
 
         store.preserveTabs(tabs, selectedTab: selectedTab)
-
+
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManager.swift:207]: nit: remove tab as we don't it it for new line

File: Client/Frontend/Browser/TabManager.swift [hunk 6]
```diff
@@ -210,6 +218,12 @@ class TabManager: NSObject {
             tab.applyTheme()
         }
     }
+    
+    func preserveTabs() {
+        print("backgrounding and preserving")
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManager.swift:223]: nit: do we need this print 🤔 ( I am guessing this is part of your cleanup )

File: Client/Frontend/Browser/TabManager.swift [hunk 7]
```diff
@@ -608,6 +622,8 @@ extension TabManager {
                 }
             }
         }
+        
+        print("restoring tabs on startup: \(count)")
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManager.swift:626]: nit: print cleanup

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 8]
```diff
@@ -31,7 +32,7 @@ class TabManagerStore {
     var hasTabsToRestoreAtStartup: Bool {
         return archivedStartupTabs.count > 0
     }
-
+
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManagerStore.swift:35]: nit: extra spacing cleanup

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 9]
```diff
@@ -95,15 +77,32 @@ class TabManagerStore {
 
         writeOperation.cancel()
 
+        // TODO: Look into this more
         let tabStateData = NSMutableData()
         let archiver = NSKeyedArchiver(forWritingWith: tabStateData)
+//        let simpleTabs = savedTabs.map { BasicSavedTab(tab: $0.toSimpleTab(), isSelected: false) }
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManagerStore.swift:83]: nit: I know this portion is a part of your cleanup but adding a comment as a reminder

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 10]
```diff
@@ -116,7 +115,9 @@ class TabManagerStore {
 
     func restoreStartupTabs(clearPrivateTabs: Bool, tabManager: TabManager) -> Tab? {
         let selectedTab = restoreTabs(savedTabs: archivedStartupTabs, clearPrivateTabs: clearPrivateTabs, tabManager: tabManager)
-        archivedStartupTabs.removeAll()
+        
+        print("restoreStartupTabs: \(archivedStartupTabs.count)")
+//        archivedStartupTabs.removeAll()
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TabManagerStore.swift:120]: nit: cleanup

File: Client/Frontend/Browser/TopSitesHandler.swift [hunk 11]
```diff
@@ -0,0 +1,65 @@
+//
+//  TopSitesHandler.swift
+//  Client
+//
+//  Created by Sawyer Blatz on 8/24/20.
+//  Copyright © 2020 Mozilla. All rights reserved.
+//
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TopSitesHandler.swift:7]: fix: Missing MPL  

```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

File: Client/Frontend/Browser/TopSitesManager.swift [hunk 12]
```diff
@@ -0,0 +1,76 @@
+//
+//  TopSitesHandler.swift
+//  Client
+//
+//  Created by Sawyer Blatz on 8/24/20.
+//  Copyright © 2020 Mozilla. All rights reserved.
+//
+
```
Comment (nbhasin2) [inline Client/Frontend/Browser/TopSitesManager.swift:8]: fix: Missing MPL

```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

File: Client/Frontend/Home/FirefoxHomeViewController.swift [hunk 13]
```diff
@@ -543,9 +542,31 @@ extension FirefoxHomeViewController: DataObserverDelegate {
...
+//                self.topSitesManager.content = Array(newSites[0..<Int(ActivityStreamTopSiteCacheSize)])
+//            } else {
+//                self.topSitesManager.content = newSites
+//            }
+//
+            self.topSitesManager.content = Array(result.prefix(maxItems))
+
+//            if newSites.count > maxItems {
+//                self.topSitesManager.content =  Array(newSites[0..<maxItems])
+//            }
```
Comment (nbhasin2) [inline Client/Frontend/Home/FirefoxHomeViewController.swift:564]: fix: cleanup

File: Client/Frontend/Strings.swift [hunk 14]
```diff
@@ -821,6 +821,10 @@ extension String {
     
     // Widget - Shared
     public static let QuickActionsGalleryTitle = NSLocalizedString("TodayWidget.QuickActionsGalleryTitle", tableName: "Today", value: "Quick Actions", comment: "Quick Actions title when widget enters edit mode")
+   
+    
+    // Quick View - Gallery View
+    public static let QuickViewGalleryTile = NSLocalizedString("TodayWidget.QuickViewGalleryTitle", tableName: "Today", value: "Quick View", comment: "Quick View title user is picking a widget to add.")
```
Comment (nbhasin2) [inline Client/Frontend/Strings.swift:827]: https://github.com/mozilla-mobile/firefox-ios/blob/main/Client/Frontend/Strings.swift#L871

I think its the same string as what we have in our main branch. Hopefully we are not adding it twice. 🤞

File: WidgetKit/Helpers.swift [hunk 15]
```diff
@@ -15,3 +17,33 @@ func linkToContainingApp(_ urlSuffix: String = "", query: String) -> URL {
...
+                    if fetchImageWork?.isCancelled == true { return }
+
+                    completion(Image(uiImage: image))
+                    fetchImageWork = nil
+                }
+            }
+        }
+    }
+
+    queue.async(execute: fetchImageWork!)
```
Comment (nbhasin2) [inline WidgetKit/Helpers.swift:39]: I think this is fine that we are using forced unwrapping but considering there is `if let` and if that fails then we could get nil. Its rare but not impossible. @sblatz thoughts on this? 

```
if let data = try? Data(contentsOf: url) {
```

File: WidgetKit/OpenTabs/OpenTabsWidget.swift [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+            $0.url?.absoluteString.starts(with: "internal://") == false &&
+            $0.title != nil
+        }
+        
+        let faviconFetchGroup = DispatchGroup()
+        
+        var tabFaviconDictionary = [String : Image]()
+        
+        
+
```
Comment (nbhasin2) [inline WidgetKit/OpenTabs/OpenTabsWidget.swift:27]: nit: cleanup

File: WidgetKit/OpenTabs/OpenTabsWidget.swift [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+import WidgetKit
+import UIKit
+import Combine
+
+struct TabProvider: TimelineProvider {
+    public typealias Entry = OpenTabsEntry
+    
+    public func snapshot(with context: Context, completion: @escaping (OpenTabsEntry) -> ()) {
+        let allOpenTabs = TabArchiver.tabsToRestore(tabsStateArchivePath: tabsStateArchivePath())
+        let openTabs = allOpenTabs.filter {
```
Comment (nbhasin2) [inline WidgetKit/OpenTabs/OpenTabsWidget.swift:15]: pretty good use of filter here 👍

File: WidgetKit/TopSites/TopSitesWidget.swift [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+import Storage
+import SyncTelemetry
+
+struct TopSitesProvider: TimelineProvider {
+    public typealias Entry = TopSitesEntry
+
+    public func snapshot(with context: Context, completion: @escaping (TopSitesEntry) -> ()) {
+        TopSitesHandler.getTopSites(profile: BrowserProfile(localName: "profile")).uponQueue(.main) { topSites in
+            
+            // TODO: It seems like some sites (Twitter, YouTube) do not have favicons... where should I fetch them? I think they're "built in"
```
Comment (nbhasin2) [inline WidgetKit/TopSites/TopSitesWidget.swift:18]: I believe we have local fevicons that we can fetch and incase they are outdated we should update them in our app.

File: WidgetKit/OpenTabs/OpenTabsWidget.swift [hunk 19]
```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import SwiftUI
+import WidgetKit
+import UIKit
+import Combine
+
+struct TabProvider: TimelineProvider {
```
Comment (nbhasin2) [inline WidgetKit/OpenTabs/OpenTabsWidget.swift:10]: Suggestion: Wondering if we can move some business logic ouf of this file to either its own file. Ex. timelines can and logic in one file and view in other along with view representation. 

What do you think of this @sblatz ?

File: package-lock.json (hunk 2 above)
Comment (sblatz) [reply package-lock.json:2818]: Agreed, think this is a random change that got pulled in.

File: WidgetKit/Helpers.swift (hunk 15 above)
Comment (sblatz) [reply WidgetKit/Helpers.swift:39]: Wrapped it in this :)

File: WidgetKit/OpenTabs/OpenTabsWidget.swift (hunk 19 above)
Comment (sblatz) [reply WidgetKit/OpenTabs/OpenTabsWidget.swift]: Done!

File: Client/Frontend/Browser/SavedTab+ConfigureExtension.swift (hunk 4 above)
Comment (sblatz) [reply Client/Frontend/Browser/SavedTab+ConfigureExtension.swift:22]: Done :)

**Comment by sblatz** [comment]
This should be ready for review again, @nishant2718. I pushed the cleanups & added the top sites "bundled" icons logic we discussed.

**Comment by nbhasin2** [comment]
Closing as top sites are already merged to main

**Pull request closed without merging**